sys/vm/vm_map.c

```diff
@@ -5154,8 +5156,11 @@ vmspace_fork(struct vmspace *vm1, vm_ooffset_t *fork_charge)
 	 * transition in progress (but an epoch might be open).
 	 *
 	 * XXX NWF We should probably go around again to force the epoch closed.
+	 * XXX-MJ I am skeptical that the state should be inherited here.
```

Comment:
It's unpleasant, I agree.  But the `pmap` is a mess between epoch opening and closing, with a mix of `CLG`s across PTEs.  So unless we either rewind time to before this epoch started (ick!) or force it to finish (also ick!), we're going to take CLG traps in the child and probably we better still think there's an epoch open. :(

This is all, ah, extremely lightly tested, since the benchmarks we ran didn't fork that often and really didn't fork that often while revocation was in progress.

In the case of async revocation we probably need to queue another work item for the child AS, though?

Reply:
I need to think about this a bit more, but we also have the option of making `pmap_copy()` ensure that the CLGs are consistent in child PTEs. I think you're right that we might need to queue a work item for the child.

Reply:
Oh, right, sorry; the PTEs aren't themselves CoW.  I, too, think I need to think more about this. :)

Reply:
I've modified this path to copy the async state to the child and conditionally queue a work item for the child vmspace. I also added a regression test which kicks off async revocation and calls fork() in the background.